Add PipeWire support to AudioRecord and AudioPlayback tests

tmoida opened the pull request:
- Updated AudioRecord and AudioPlayback scripts to support both PulseAudio and PipeWire
- Added AUDIO_BACKEND variable for backend selection
- Updated README files with usage instructions for both backends
- Modified LAVA test plan to run tests with both PulseAudio and PipeWire

smuppand reviewed (changes requested):
@tmoida Could you please squash the commits into a single one?

smuppand commented on Runner/suites/Multimedia/Audio/AudioPlayback/run.sh:
```diff
@@ -29,26 +29,43 @@ fi
+        TESTBINARY="paplay"
+        PLAY_CMD="paplay \"$PLAYBACK_CLIP\" -d \"$AUDIO_DEVICE\""
+        ;;
+    pipewire)
```
Great to expose timeouts/ volume/ loops via env. Add a comment for a valid value range (eg paply volume 0-65536, pw-play 0.0 - etc)

smuppand commented:
@tmoida , please squash your commits.

smuppand commented on Runner/suites/Multimedia/Audio/AudioPlayback/run.sh:
```diff
@@ -68,16 +108,19 @@ fi
 
 # --- Start the Playback, capture output ---
-timeout 15s paplay "$PLAYBACK_CLIP" -d "$AUDIO_DEVICE" > "$LOGDIR/playback_stdout.log" 2>&1
+for i in $(seq 1 "$PLAYBACK_LOOPS"); do
```
Could you review the shellint-reported errors regarding unused variables? https://github.com/qualcomm-linux/qcom-linux-testkit/actions/runs/16964336815/job/48455697686?pr=137

tmoida commented:
Hi @smuppand,

I've squashed the commits and fixed the Shell Check warning by using the loop variable. Please let me know if there is anything else that needs to be addressed.

Also. I have validated these scripts on Yocto builds, and they work fine. However, on Mainline (Nightly sanity) builds, they are being skipped due to a missing timeout binary. I wanted to check: will the timeout binary be included in future builds, or should I update the script to avoid using it?

smuppand commented:
> Hi @smuppand,
> 
> I've squashed the commits and fixed the Shell Check warning by using the loop variable. Please let me know if there is anything else that needs to be addressed.
> 
> Also. I have validated these scripts on Yocto builds, and they work fine. However, on Mainline (Nightly sanity) builds, they are being skipped due to a missing timeout binary. I wanted to check: will the timeout binary be included in future builds, or should I update the script to avoid using it?

You can use the library function run_with_timeout() if the build does not include the timeout utility.

tmoida commented:
Hi @smuppand,

Thank you for your feedback.

I have implemented the requested modular structure and made the following changes:

```
1. Moved all core logic for playback and recording into a shared `libaudio.sh`.
2. Created minimal `run.sh` scripts under AudioPlayback and AudioRecord that only parse CLI flags and invoke functions from `libaudio.sh`.
3. Wrapped all binary calls (paplay, pw-play, parec, pw-record) inside `libaudio.sh`, eliminating direct calls in `run.sh`.
4. Added support for both CLI flags and environment variables for configuration.
```

**Why separate run.sh scripts for playback and record?**
Regarding the use of separate `run.sh` scripts for playback and recording: We are utilizing `run-test.sh` as the main test runner, which expects each test case to be in its own directory and produce a result file named after that directory (e.g., `AudioPlayback.res`, `AudioRecord.res`). Using a single `run.sh` within a common Audio/ directory would prevent `run-test.sh` from locating the correct `.res` file, potentially leading to false negatives, even when tests pass.

To address this and ensure compatibility with `run-test.sh`, we have maintained separate directories and minimal `run.sh` wrappers for each test case. This approach allows for correct result parsing while centralizing the logic in `libaudio.sh`.

All usage details, CLI options, and examples are documented in the respective `Read_me.md` files for easy reference.

Let me know if you’d like any further refinements!

smuppand commented on Runner/suites/Multimedia/Audio/AudioPlayback/run.sh:
```diff
@@ -1,94 +1,66 @@
+  SEARCH=$(dirname "$SEARCH")
 done
+[ -z "$INIT_ENV" ] && echo "[ERROR] init_env not found" && exit 1
+# shellcheck source=/dev/null
```
using `# shellcheck source=/dev/null` is a blunt silencer that hides real problems and degrades analysis. It tells ShellCheck “pretend this sourced file exists but is empty, It masks real bugs.

smuppand commented on Runner/suites/Multimedia/Audio/AudioPlayback/run.sh:
```diff
@@ -1,94 +1,66 @@
+. "$TOOLS/libaudio.sh"
 
-test_path=$(find_test_case_by_name "$TESTNAME")
-cd "$test_path" || exit 1
```
This approach will fail if we run any test from the run-test root, and the .res file will be generated in the directory where the command is executed, rather than in the test directory. Please keep the expected snippet.

smuppand commented on Runner/suites/Multimedia/Audio/AudioPlayback/run.sh:
```diff
@@ -1,94 +1,66 @@
-test_path=$(find_test_case_by_name "$TESTNAME")
-cd "$test_path" || exit 1
+# Override with BusyBox-compatible version
+check_network_status() {
```
check_network_status override isn’t used—drop it to keep things tight. 
Use available built-in fiunctions like [get_ip_address()](https://github.com/qualcomm-linux/qcom-linux-testkit/blob/main/Runner/utils/functestlib.sh#L498) to avoid duplication.

smuppand commented on Runner/suites/Multimedia/Audio/AudioRecord/run.sh:
```diff
@@ -1,77 +1,62 @@
+  SEARCH=$(dirname "$SEARCH")
 done
+[ -z "$INIT_ENV" ] && echo "[ERROR] init_env not found" && exit 1
+# shellcheck source=/dev/null
```
NIT

smuppand commented on Runner/suites/Multimedia/Audio/AudioRecord/run.sh:
```diff
@@ -1,77 +1,62 @@
-    # shellcheck disable=SC1090
-    . "$INIT_ENV"
-fi
-# shellcheck disable=SC1090,SC1091
```
NIT

smuppand commented on Runner/suites/Multimedia/Audio/AudioRecord/run.sh:
```diff
@@ -1,77 +1,62 @@
-RESULT_FILE="$TESTNAME.res"
 
-test_path=$(find_test_case_by_name "$TESTNAME")
-cd "$test_path" || exit 1
```
NIT

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+    return 0
+}
+
+execute_with_timeout() {
```
why to introduce a new function instead of using run_with_timeout?

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+}
+
+audio_download_clip() {
+    TAR_URL="https://github.com/qualcomm-linux/qcom-linux-testkit/releases/download/Pulse-Audio-Files-v1.0/AudioClips.tar.gz"
```
this should go to a file constant

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+    done
+
+    get_kernel_log > "$LOGDIR/dmesg_after.log"
+    scan_dmesg_errors "audio" "$LOGDIR"
```
Your function signature is scan_dmesg_errors(prefix, module_regex, [exclude]).
 
Current calls use scan_dmesg_errors "audio" "$LOGDIR" (reversed). This silently weakens the scan. It returns 0 when errors are found, 1 when clean. Runners should only gate FAIL on this if a strict flag is set (to avoid flakiness).

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+
+    for i in $(seq 1 "$PLAYBACK_LOOPS"); do
+        log_info "Playback loop $i of $PLAYBACK_LOOPS"
+        execute_with_timeout "$PLAYBACK_TIMEOUT" sh -c "$PLAY_CMD" >> "$LOGDIR/playback_stdout.log" 2>&1
```
Defaults like 15s are fine for timeout(1) but may break run_with_timeout (usually expects integers). Normalize once, inside execute_with_timeout.

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+
+    get_kernel_log > "$LOGDIR/dmesg_before.log"
+    PLAY_CMD=$(setup_playback_command)
+    PLAY_SUCCESS=0
```
Playback uses PLAY_SUCCESS but sets PLAYBACK_SUCCESS in one branch—typo causes false PASS.

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+    if [ "$AUDIO_BACKEND" = "pipewire" ]; then
+        select_pipewire_source
+    fi
+    check_dependencies "$([ "$AUDIO_BACKEND" = "pulseaudio" ] && echo "parec" || echo "pw-record")" pgrep grep
```
select_pipewire_source uses wpctl but dependency check doesn’t require it.
 
Minor: pgrep grep in check_dependencies is unnecessary; keep pgrep, drop grep.

smuppand commented on Runner/utils/libaudio.sh:
```diff
@@ -0,0 +1,200 @@
+        log_fail_exit "$TESTNAME" "Audio playback test failed"
+        echo "$TESTNAME FAIL" > "$RESULT_FILE"
+    fi
+
```
Transferring PASS/FAIL/SKIP ownership to the top-level runner. These shared libraries will be used across the audio files.

smuppand reviewed (changes requested):
Also woth adding 

- WAV 16-bit/48k, 44.1k; Ogg Vorbis/Opus (if tools available), AAC.
- Short (5s), medium (30s), long (120s) smoke tiers.
- Mixer dumps (pactl list, pw-dump) on failure.

smuppand commented on Runner/utils/functestlib.sh:
```diff
@@ -263,7 +263,7 @@ check_network_status() {
 
     # Get first active IPv4 address (excluding loopback)
-    ip_addr=$(ip -4 addr show scope global up | grep -oP '(?<=inet\s)\d+(\.\d+){3}' | head -n 1)
+	ip_addr=$(ip -4 addr show scope global up | awk '/inet / {print $2}' | cut -d/ -f1 | head -n 1)
```
AFAIK, this approach won't work if the routing table includes multiple NICs, VRFs, or asymmetric configurations. It might be better to add an interface-scan method to simply retrieve the first global address.

smuppand commented on Runner/utils/audio_common.sh:
```diff
@@ -0,0 +1,548 @@
+#!/bin/sh
+# SPDX-License-Identifier: BSD-3-Clause-Clear
```
The copyright information is missing.

smuppand commented on Runner/utils/audio_common.sh:
```diff
@@ -0,0 +1,551 @@
+
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
```
Remove white space.

smuppand reviewed (approved):
LGTM
